src/kaleidoscope/Device.h

```diff
@@ -0,0 +1,502 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file kaleidoscope/Hardware.h
+ * Base class for Kaleidoscope hardware libraries.
+ */
+
+#pragma once
+
+#include "kaleidoscope/MatrixAddr.h"
+#include "kaleidoscope_internal/deprecations.h"
+#include "kaleidoscope/macro_helpers.h"
+
+#include "kaleidoscope/driver/keyscanner/None.h"
+#include "kaleidoscope/driver/leddriver/None.h"
+#include "kaleidoscope/driver/mcu/None.h"
+#include "kaleidoscope/driver/bootloader/None.h"
+#include "kaleidoscope/driver/storage/None.h"
+
+#ifndef CRGB
+#error cRGB and CRGB *must* be defined before including this header!
+#endif
+
+#define _DEPRECATED_MESSAGE_HARDWARE_RESETDEVICE                       \
+  "`KeyboardHardware.resetDevice()` is deprecated, please use " __NL__ \
+  "`KeyboardHardware.rebootBootloader()` instead."
+
+/* All hardware libraries must define the following macros:
+ * HARDWARE_IMPLEMENTATION - the name of your public object conforming to
+ *   the 'class Hardware' interface below.
+ * CRGB(r,g,b) - explained below
+ * cRGB, a structure with at least three members: r, g, and b -
+ * compilation will fail otherwise.
+ *
+ * Despite its name, the members do not need to be in the order r g b -- most
+ * likely they will be in an order that is convenient for the hardware. So
+ * initializing a cRGB with a struct literal will give surprising results for any
+ * colors where r, g, and b do not have the same value. Each Hardware library
+ * defines a CRGB(r,g,b) macro which returns a literal cRGB with the given values.
+ */
+
+namespace kaleidoscope {
+
+struct DeviceProps {
+  typedef kaleidoscope::driver::keyscanner::NoProps KeyScannerProps;
+  typedef kaleidoscope::driver::keyscanner::None KeyScanner;
+  typedef kaleidoscope::driver::leddriver::NoProps LEDDriverProps;
+  typedef kaleidoscope::driver::leddriver::None LEDDriver;
+  typedef kaleidoscope::driver::mcu::None MCU;
+  typedef kaleidoscope::driver::bootloader::None Bootloader;
+  typedef kaleidoscope::driver::storage::NoProps StorageProps;
+  typedef kaleidoscope::driver::storage::None Storage;
+};
+
+template<typename _DeviceProps>
+class Device {
+ private:
+  class DummySerial {
+   public:
+    DummySerial() {}
+  };
+
+  static DummySerial dummy_serial_;
+
+ public:
+
+  typedef _DeviceProps Props;
+
+  typedef typename _DeviceProps::KeyScanner KeyScanner;
+  typedef typename _DeviceProps::KeyScannerProps KeyScannerProps;
+  typedef typename _DeviceProps::KeyScannerProps::KeyAddr KeyAddr;
+  typedef typename _DeviceProps::LEDDriverProps LEDDriverProps;
+  typedef typename _DeviceProps::LEDDriver LEDDriver;
+  typedef typename _DeviceProps::MCU MCU;
```

Comment:
Some drivers appear to be missing props-typedefs here.

Reply:
Yeah, MCU and Bootloader have no props, because none made sense yet.